infra/azure/modules/auth/main.tf

```diff
@@ -58,12 +58,53 @@ resource "azuread_application_password" "oauth2" {
   application_object_id = azuread_application.oauth2.object_id
 }
 
+resource "random_uuid" "hailctl_oauth2_idenfier_uri_id" {}
+resource "random_uuid" "hailctl_oauth2_scope_id" {}
+
+resource "azuread_application" "hailctl_oauth2" {
+  display_name = "${var.resource_group_name}-hailctl-oauth2"
+
+  identifier_uris = ["api://hail-${random_uuid.hailctl_oauth2_idenfier_uri_id.result}"]
+
+  public_client {
+    redirect_uris = ["http://localhost/"]
+  }
+
+  api {
+    oauth2_permission_scope {
+      admin_consent_description  = "Allow the hail python library to access the Hail Batch service on behalf of the signed-in user."
+      admin_consent_display_name = "hailctl"
+      user_consent_description   = "Allow the hail python library to access the Hail Batch service on your behalf."
+      user_consent_display_name  = "hailctl"
+      enabled                    = true
+      id                         = random_uuid.hailctl_oauth2_scope_id.result
+      type                       = "User"
+      value                      = "batch.default"
+    }
+  }
+}
+
 locals {
   oauth2_credentials = {
     appId    = azuread_application.oauth2.application_id
     password = azuread_application_password.oauth2.value
     tenant   = data.azurerm_client_config.primary.tenant_id
   }
+
+  appIdentifierUri = "api://hail-${random_uuid.hailctl_oauth2_idenfier_uri_id.result}"
+  userOauthScope   = "${local.appIdentifierUri}/batch.default"
+  spOauthScope     = "${local.appIdentifierUri}/.default"
+
+  hailctl_oauth2_credentials = {
+    appId            = azuread_application.hailctl_oauth2.application_id
+    appIdentifierUri = local.appIdentifierUri
+    # For some reason SP client secret authentication refuses scopes that are not .default and this returned a valid token with
+    # the desired audience. When creating the oauth scope, terraform refused to create a scope that started with a `.` e.g. `.default`, and
+    # as such was forced to create the scope `batch.default`. Whether this is a bug in the terraform provider or a feature of AAD is unclear.
```

Comment:
Any compelling reason to not start with just `.default` for everyone?

Reply:
Not sure I get what you mean. I tried using just `.default` and terraform called it invalid.

Reply:
I also don't recall what I meant.